File: xswap/xswap.py
from typing import List, Set, Tuple

import xswap._xswap_backend


def permute_edge_list(edge_list: List[Tuple[int, int]], allow_self_loops: bool = False,
                      allow_antiparallel: bool = False, multiplier: float = 10,
                      excluded_edges: Set[Tuple[int, int]] = set(), seed: int = 0,
                      max_malloc: int = 400000_000):
    """
    Permute the edges of a graph using the XSwap method given by Hanhijärvi,
    et al. (doi.org/f3mn58). XSwap is a degree-preserving network randomization
    technique that selects edges, checks the validity of the swap, and exchanges
    the target nodes between the edges. For information on what values to select
    for directed, please see README.md.

    Parameters
    ----------
    edge_list : List[Tuple[int, int]]
        Edge list representing the graph to be randomized. Tuples can contain
        integer values representing nodes. No value should be greater than C++'s
        `INT_MAX`, in this case 2_147_483_647.
    allow_self_loops : bool
        Whether to allow edges like (0, 0). In the case of bipartite graphs,
        such an edge represents a connection between two distinct nodes, while
        in other graphs it may represent an edge from a node to itself, in which
        case an edge may or may not be meaningful depending on context.
    allow_antiparallel : bool
        Whether to allow simultaneous edges like (0, 1) and (1, 0). In the case
        of bipartite graphs, these edges represent two connections between four
        distinct nodes, while for other graphs, these may be connections between
        the same two nodes.
    multiplier : float
        The number of edge swap attempts is determined by the product of the
        number of existing edges and multiplier. For example, if five edges are
        passed and multiplier is set to 10, 50 swaps will be attempted. Non-integer
        products will be rounded down to the nearest integer.
    excluded_edges : Set[Tuple[int, int]]
        Specific edges which should never be created by the network randomization
    seed : int
        Random seed that will be passed to the C++ Mersenne Twister 19937 random
        number generator.
    max_malloc : int (`unsigned long long int` in C)
        The maximum amount of memory to be allocated using `malloc` when making
        a bitset to hold edges. An uncompressed bitset is implemented for
        holding edges that is significantly faster than alternatives. However,
        it is memory-inefficient and will not be used if more memory is required
        than `max_malloc`. Above the threshold, a Roaring bitset will be used.

    Returns
    -------
    new_edges : List[Tuple[int, int]]
        Edge list of a permutation of the network given as `edge_list`
    stats : Dict[str, int]
        Information about the permutation performed. Gives the following information:
        `swap_attempts` - number of attempted swaps
        `same_edge` - number of swaps rejected because one edge was chosen twice
        `self_loop` - number of swaps rejected because new edge is a self-loop
        'duplicate` - number of swaps rejected because new edge already exists
        `undir_duplicate` - number of swaps rejected because the network is
            undirected and the reverse of the new edge already exists
        `excluded` - number of swaps rejected because new edge was among excluded
    """
    if len(edge_list) != len(set(edge_list)):
        raise ValueError("Edge list contained duplicate edges.")

    # Number of attempted XSwap swaps
    num_swaps = int(multiplier * len(edge_list))

    # Compute the maximum node ID (for creating the bitset)
    max_id = max(map(max, edge_list))

    new_edges, stats = xswap._xswap_backend._xswap(
        edge_list, list(excluded_edges), max_id, allow_self_loops,
        allow_antiparallel, num_swaps, seed, max_malloc)

    return new_edges, stats

File: xswap/xswap.cpp
#include <random>
#include "xswap.h"

void swap_edges(int **edges, int num_edges, int **excluded_edges, int num_excluded_edges,
                 EdgeHashTable edges_set, int seed, int directed, int num_swaps,
                 int *same_edge, int *self_loop, int *duplicate,int *undir_duplicate,
                 int *excluded) {
    // Initialize unbiased random number generator
    std::mt19937 rng(seed);
    std::uniform_int_distribution<int> uni(0, num_edges - 1);

    // Do XSwap
    for (int i = 0; i < num_swaps; i++) {
        // Draw edges randomly
        int edge_index_a = uni(rng);
        int edge_index_b = uni(rng);

        if (edge_index_a == edge_index_b) {
            same_edge += 1;
            continue;
        }

        // Old edges
        int* edge_a = edges[edge_index_a];
        int* edge_b = edges[edge_index_b];

        // Form potential new edges
        int new_edge_a[2] = { edge_a[0], edge_b[1] };
        int new_edge_b[2] = { edge_b[0], edge_a[1] };
        int* new_edges[2] = { new_edge_a, new_edge_b };
        // Check validity of both new edges
        bool valid = true;
        for (int i = 0; i < 2; i++) {
            int* new_edge = new_edges[i];
            // New edge would be a self-loop
            if (new_edge[0] == new_edge[1]) {
                self_loop += 1;
                valid = false;
                break;
            }
            // New edge already exists
            if (edges_set.contains(new_edge)) {
                duplicate += 1;
                valid = false;
                break;
            }
            // Undirected and reverse of new edge already exists
            int reversed[2] = { new_edge[1], new_edge[0] };
            if (!directed && edges_set.contains(reversed)) {
                undir_duplicate += 1;
                valid = false;
                break;
            }
            // New edges are in excluded edges. Because few excluded edges are expected,
            // this iterates in O(n) rather than allocating a large hash table for all
            // possible edges. Considerable memory savings when excluding few edges.
            // An alternative is to pass excluded edges as a Python set and to use the
            // built-in PySet_Contains function to check membership. This has the downside
            // that the edge must be instantiated as a new Python tuple object and the
            // values represented as new Python long integer objects.
            for (int i = 0; i < num_excluded_edges; i++) {
                if (excluded_edges[i][0] == new_edge[0] && excluded_edges[i][1] == new_edge[1]) {
                    excluded += 1;
                    valid = false;
                    break;
                }
            }
        }
        if (valid) {
            edges_set.remove(edge_a);
            edges_set.remove(edge_b);

            int temp_target = edge_a[1];
            edge_a[1] = edge_b[1];
            edge_b[1] = temp_target;

            edges_set.add(edge_a);
            edges_set.add(edge_b);
        }
    }
}
